File: skills/docforge/scripts/query_catalog.js
#!/usr/bin/env node
"use strict";
/** Query the split Docforge catalog. Agents and scripts use this — not raw files. */

const fs = require("fs");
const path = require("path");
const { dumpJson, fail } = require("./_util");

const SKILL_ROOT = path.resolve(__dirname, "..");
const CATALOG_DIR = path.join(SKILL_ROOT, ".metadata", "catalog");
const INDEX_PATH = path.join(CATALOG_DIR, "index.json");
const TYPES_DIR = path.join(CATALOG_DIR, "types");
const PROFILES_DIR = path.join(CATALOG_DIR, "profiles");
const CONTRACTS_DIR = path.join(SKILL_ROOT, "references", "catalog-contracts");
const PROFILE_DIMENSIONS = ["shapes", "platforms", "frameworks", "concerns", "audiences"];
const ALLOWED_DOMINANT_FORMS = new Set([
  null,
  undefined,
  "table",
  "flowchart",
  "sequenceDiagram",
  "erDiagram",
]);
const REQUIRED_DOC_FIELDS = [
  "id",
  "type",
  "path",
  "group",
  "selection",
  "scaffold_template",
  "requires",
  "target_depth",
  "write_order",
  "provenance_mode",
  "audit_profile",
];

function loadIndex() {
  if (!fs.existsSync(INDEX_PATH)) throw new Error(`catalog index not found: ${INDEX_PATH}`);
  return JSON.parse(fs.readFileSync(INDEX_PATH, "utf8"));
}

function loadProfile(dimension) {
  const aliases = {
    shape: "shapes",
    platform: "platforms",
    framework: "frameworks",
    concern: "concerns",
    audience: "audiences",
  };
  dimension = aliases[dimension] || dimension;
  if (!PROFILE_DIMENSIONS.includes(dimension)) {
    throw new Error(
      `unknown profile dimension: ${dimension}; expected one of: ${PROFILE_DIMENSIONS.join(", ")}`,
    );
  }
  const target = path.join(PROFILES_DIR, `${dimension}.json`);
  if (!fs.existsSync(target)) throw new Error(`profile file not found: ${target}`);
  return JSON.parse(fs.readFileSync(target, "utf8"));
}

function loadProfiles() {
  const profiles = {};
  for (const dimension of PROFILE_DIMENSIONS) profiles[dimension] = loadProfile(dimension);
  return profiles;
}

function loadType(docId) {
  const target = path.join(TYPES_DIR, `${docId}.json`);
  if (!fs.existsSync(target)) throw new Error(`unknown document type id: ${docId}`);
  return JSON.parse(fs.readFileSync(target, "utf8"));
}

function loadAllTypes() {
  return loadIndex().document_types.map((row) => loadType(row.id));
}

function asLegacyCatalog() {
  const index = loadIndex();
  const tiers = index.tiers;
  const tierList = Array.isArray(tiers)
    ? tiers
    : Object.entries(tiers).map(([id, meta]) => ({ id, order: meta.order }));
  return {
    $schema: "catalog-schema.json",
    version: index.version,
    tiers: tierList,
    profiles: loadProfiles(),
    groups: index.groups || [],
    capabilities: index.capabilities || [],
    documents: loadAllTypes(),
    cue_hints: index.cue_hints || [],
  };
}

function tierRows(tier) {
  const index = loadIndex();
  const tiers = index.tiers;
  const known = Array.isArray(tiers) ? new Set(tiers.map((t) => t.id)) : new Set(Object.keys(tiers));
  if (!known.has(tier)) throw new Error(`unknown tier: ${tier}`);
  return index.document_types.filter((row) => row.tier === tier);
}

function mergedRecord(docId) {
  const index = loadIndex();
  const row = index.document_types.find((r) => r.id === docId);
  if (!row) throw new Error(`unknown document type id: ${docId}`);
  return { ...loadType(docId), tier: row.tier, index_path: row.path };
}

function normalizeProfileIds(dimension, values, profiles) {
  const aliases = {};
  for (const definition of profiles[dimension]) {
    aliases[definition.id] = definition.id;
    for (const alias of definition.aliases || []) aliases[alias] = definition.id;
  }
  const resolved = [];
  for (const value of values) {
    if (!(value in aliases)) throw new Error(`unknown ${dimension.slice(0, -1)}: ${value}`);
    const canonical = aliases[value];
    if (!resolved.includes(canonical)) resolved.push(canonical);
  }
  return resolved;
}

function applicable(options = {}) {
  const index = loadIndex();
  const profiles = loadProfiles();
  const selected = {
    shapes: normalizeProfileIds("shapes", options.shape || [], profiles),
    platforms: normalizeProfileIds("platforms", options.platform || [], profiles),
    frameworks: normalizeProfileIds("frameworks", options.framework || [], profiles),
    concerns: normalizeProfileIds("concerns", options.concern || [], profiles),
    audiences: normalizeProfileIds("audiences", options.audience || [], profiles),
  };
  const tiers = index.tiers;
  const ranks = Array.isArray(tiers)
    ? Object.fromEntries(tiers.map((t) => [t.id, t.order]))
    : Object.fromEntries(Object.entries(tiers).map(([id, meta]) => [id, meta.order]));
  const tier = options.tier || "diligence";
  if (!(tier in ranks)) throw new Error(`unknown tier: ${tier}`);
  const tierRank = ranks[tier];
  const ids = [];
  for (const row of index.document_types) {
    const detail = loadType(row.id);
    const rule = detail.selection;
    if (rule.mode === "dynamic" && !options.includeDynamic) continue;
    if (ranks[rule.min_tier] > tierRank) continue;
    const selectors = rule.selectors || {};
    const hasSelectors = Object.values(selectors).some((v) => v && v.length);
    if (hasSelectors) {
      let matched = false;
      for (const [dimension, values] of Object.entries(selectors)) {
        if (dimension === "frameworks") continue;
        const selectedValues = selected[dimension] || [];
        if (values.some((value) => selectedValues.includes(value))) {
          matched = true;
          break;
        }
      }
      if (!matched) continue;
    }
    ids.push(detail.id);
  }
  return ids;
}

function validate() {
  const errors = [];
  let index;
  try {
    index = loadIndex();
  } catch (error) {
    return [error.message];
  }
  if (index.version !== "2.4.0") {
    errors.push(`catalog version must be 2.4.0, got ${index.version}`);
  }
  for (const key of ["tiers", "groups", "capabilities", "document_types"]) {
    if (!(key in index)) errors.push(`index.json missing ${key}`);
  }
  const tiers = index.tiers || {};
  const tierIds = Array.isArray(tiers) ? new Set(tiers.map((t) => t.id)) : new Set(Object.keys(tiers));

  const profiles = {};
  const profileIds = {};
  for (const dimension of PROFILE_DIMENSIONS) {
    const target = path.join(PROFILES_DIR, `${dimension}.json`);
    if (!fs.existsSync(target)) {
      errors.push(`missing profile file: ${path.basename(target)}`);
      profiles[dimension] = [];
      profileIds[dimension] = new Set();
      continue;
    }
    const definitions = JSON.parse(fs.readFileSync(target, "utf8"));
    profiles[dimension] = definitions;
    const ids = new Set(definitions.map((item) => item.id));
    if (ids.size !== definitions.length) errors.push(`${dimension}: duplicate profile id`);
    profileIds[dimension] = ids;
    const names = {};
    for (const item of definitions) {
      for (const name of [item.id, ...(item.aliases || [])]) {
        if (typeof name !== "string" || !/^[a-z0-9][a-z0-9-]*$/.test(name)) {
          errors.push(`${dimension}: invalid profile name ${name}`);
          continue;
        }
        if (name in names) {
          errors.push(
            `${dimension}: profile name collision ${name} between ${names[name]} and ${item.id}`,
          );
        }
        names[name] = item.id;
      }
    }
  }

  const indexIds = (index.document_types || []).map((row) => row.id);
  if (new Set(indexIds).size !== indexIds.length) {
    errors.push("index.json has duplicate document type ids");
  }
  const typeFiles = new Set(
    fs.readdirSync(TYPES_DIR).filter((name) => name.endsWith(".json")).map((name) => name.slice(0, -5)),
  );
  for (const docId of indexIds) {
    if (!typeFiles.has(docId)) errors.push(`index references missing type file: ${docId}.json`);
  }
  for (const stem of [...typeFiles].sort()) {
    if (!indexIds.includes(stem)) errors.push(`orphan type file not in index: ${stem}.json`);
  }

  const groups = new Set(index.groups || []);
  const capabilities = new Set(index.capabilities || []);
  const staticIds = new Set();
  const staticPaths = new Set();
  const dynamicTypes = new Set();
  let contractTypes = new Set();
  if (fs.existsSync(CONTRACTS_DIR)) {
    contractTypes = new Set(
      fs
        .readdirSync(CONTRACTS_DIR)
        .filter((name) => name.endsWith(".md") && name !== "README.md")
        .map((name) => name.slice(0, -3)),
    );
  }

  for (const docId of indexIds) {
    let doc;
    try {
      doc = loadType(docId);
    } catch (error) {
      errors.push(error.message);
      continue;
    }
    const missing = REQUIRED_DOC_FIELDS.filter((field) => !(field in doc));
    if (missing.length) {
      errors.push(`${docId}: missing fields: ${missing.join(", ")}`);
      continue;
    }
    if (doc.id !== docId) errors.push(`${docId}: id field mismatch (${doc.id})`);
    const selection = doc.selection || {};
    if (!groups.has(doc.group)) errors.push(`${docId}: unknown group ${doc.group}`);
    if (!tierIds.has(selection.min_tier)) {
      errors.push(`${docId}: unknown tier ${selection.min_tier}`);
    }
    const selectors = selection.selectors || {};
    if (selectors.frameworks && selectors.frameworks.length) {
      errors.push(`${docId}: frameworks may tailor evidence but must not select documents`);
    }
    for (const [dimension, values] of Object.entries(selectors)) {
      if (!(dimension in profileIds)) {
        errors.push(`${docId}: unknown selector dimension ${dimension}`);
        continue;
      }
      for (const value of values) {
        if (!profileIds[dimension].has(value)) {
          errors.push(`${docId}: unknown ${dimension} selector ${value}`);
        }
      }
    }
    for (const requirement of doc.requires || []) {
      if (!capabilities.has(requirement)) {
        errors.push(`${docId}: unknown requirement ${requirement}`);
      }
    }
    if (!Number.isInteger(doc.write_order)) {
      errors.push(`${docId}: write_order must be an integer`);
    }
    if (!ALLOWED_DOMINANT_FORMS.has(doc.dominant_form)) {
      errors.push(`${docId}: invalid dominant_form ${JSON.stringify(doc.dominant_form)}`);
    }
    if (contractTypes.size && !contractTypes.has(doc.type)) {
      errors.push(`${docId}: document type missing from catalog-contracts/`);
    }
    const template = path.join(SKILL_ROOT, "assets", "templates", doc.scaffold_template);
    if (!fs.existsSync(template)) {
      errors.push(`${docId}: missing template ${doc.scaffold_template}`);
    }
    if (doc.instruction_file) {
      const instruction = path.join(SKILL_ROOT, "instructions", doc.instruction_file);
      if (!fs.existsSync(instruction)) {
        errors.push(`${docId}: missing instruction ${doc.instruction_file}`);
      }
    }
    if (selection.mode === "static") {
      if (staticIds.has(doc.id)) errors.push(`duplicate static id: ${doc.id}`);
      if (staticPaths.has(doc.path)) errors.push(`duplicate static path: ${doc.path}`);
      staticIds.add(doc.id);
      staticPaths.add(doc.path);
    } else if (selection.mode === "dynamic") {
      if (dynamicTypes.has(doc.type)) errors.push(`duplicate dynamic type: ${doc.type}`);
      dynamicTypes.add(doc.type);
    } else {
      errors.push(`${docId}: selection.mode must be static or dynamic`);
    }
  }

  const infra = (profiles.shapes || []).find((item) => item.id === "infrastructure-platform");
  if (!infra) {
    errors.push("shapes: infrastructure-platform missing");
  } else {
    const patterns = new Set((infra.signals || []).map((s) => s.pattern));
    for (const required of ["ansible.cfg", "**/kustomization.yaml", "**/playbook*.yml"]) {
      if (!patterns.has(required)) {
        errors.push(`infrastructure-platform missing signal pattern ${required}`);
      }
    }
    const aliases = new Set(infra.aliases || []);
    for (const alias of ["deployment-config", "iac"]) {
      if (!aliases.has(alias)) {
        errors.push(`infrastructure-platform missing alias ${alias}`);
      }
    }
  }
  return errors;
}

function parseArgs(argv) {
  const args = {
    shape: [],
    platform: [],
    framework: [],
    concern: [],
    audience: [],
    applicableTier: "diligence",
    includeDynamic: false,
  };
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    const next = () => {
      i += 1;
      return argv[i];
    };
    if (arg === "--tier") args.tier = next();
    else if (arg === "--id") args.docId = next();
    else if (arg === "--ids") args.ids = next();
    else if (arg === "--profile") args.profile = next();
    else if (arg === "--applicable") args.applicable = true;
    else if (arg === "--shape") args.shape.push(next());
    else if (arg === "--platform") args.platform.push(next());
    else if (arg === "--framework") args.framework.push(next());
    else if (arg === "--concern") args.concern.push(next());
    else if (arg === "--audience") args.audience.push(next());
    else if (arg === "--applicable-tier") args.applicableTier = next();
    else if (arg === "--include-dynamic") args.includeDynamic = true;
    else if (arg === "--legacy") args.legacy = true;
    else if (arg === "--validate") args.validate = true;
    else throw new Error(`unknown flag: ${arg}`);
  }
  return args;
}

function main(argv = process.argv.slice(2)) {
  let args;
  try {
    args = parseArgs(argv);
  } catch (error) {
    return fail(error.message, 2);
  }
  const modes = [
    args.tier,
    args.docId,
    args.ids,
    args.profile,
    args.applicable,
    args.validate,
    args.legacy,
  ].filter(Boolean).length;
  if (modes !== 1) {
    return fail(
      "specify exactly one of --tier, --id, --ids, --profile, --applicable, --legacy, --validate",
      2,
    );
  }
  try {
    if (args.validate) {
      const errors = validate();
      if (errors.length) {
        for (const error of errors) process.stderr.write(`error: ${error}\n`);
        process.stderr.write(`${errors.length} validation error(s)\n`);
        return 1;
      }
      process.stdout.write("catalog ok\n");
      return 0;
    }
    if (args.legacy) {
      process.stdout.write(dumpJson(asLegacyCatalog()));
      return 0;
    }
    if (args.tier) {
      process.stdout.write(dumpJson(tierRows(args.tier)));
      return 0;
    }
    if (args.docId) {
      process.stdout.write(dumpJson(mergedRecord(args.docId)));
      return 0;
    }
    if (args.ids) {
      const ids = args.ids.split(",").map((part) => part.trim()).filter(Boolean);
      process.stdout.write(dumpJson(ids.map((docId) => mergedRecord(docId))));
      return 0;
    }
    if (args.profile) {
      process.stdout.write(dumpJson(loadProfile(args.profile)));
      return 0;
    }
    if (args.applicable) {
      process.stdout.write(
        dumpJson(
          applicable({
            shape: args.shape,
            platform: args.platform,
            framework: args.framework,
            concern: args.concern,
            audience: args.audience,
            tier: args.applicableTier,
            includeDynamic: args.includeDynamic,
          }),
        ),
      );
      return 0;
    }
  } catch (error) {
    return fail(error.message, 2);
  }
  return 2;
}

module.exports = {
  loadIndex,
  loadProfile,
  loadProfiles,
  loadType,
  loadAllTypes,
  asLegacyCatalog,
  tierRows,
  mergedRecord,
  applicable,
  validate,
  main,
};

if (require.main === module) {
  process.exitCode = main();
}

File: skills/docforge/scripts/query_catalog.py
#!/usr/bin/env python3
"""Query the split Docforge catalog. Agents and scripts use this — not raw files."""

from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path

from _util import dump_json, fail

SKILL_ROOT = Path(__file__).resolve().parent.parent
CATALOG_DIR = SKILL_ROOT / ".metadata" / "catalog"
INDEX_PATH = CATALOG_DIR / "index.json"
TYPES_DIR = CATALOG_DIR / "types"
PROFILES_DIR = CATALOG_DIR / "profiles"
CONTRACTS_DIR = SKILL_ROOT / "references" / "catalog-contracts"
ALLOWED_DOMINANT_FORMS = {
    None,
    "table",
    "flowchart",
    "sequenceDiagram",
    "erDiagram",
}
PROFILE_DIMENSIONS = ["shapes", "platforms", "frameworks", "concerns", "audiences"]
REQUIRED_DOC_FIELDS = {
    "id",
    "type",
    "path",
    "group",
    "selection",
    "scaffold_template",
    "requires",
    "target_depth",
    "write_order",
    "provenance_mode",
    "audit_profile",
}


def load_index() -> dict:
    if not INDEX_PATH.is_file():
        raise ValueError(f"catalog index not found: {INDEX_PATH}")
    return json.loads(INDEX_PATH.read_text(encoding="utf-8"))


def load_profile(dimension: str) -> list:
    # Accept singular aliases from CLI: shape → shapes
    aliases = {
        "shape": "shapes",
        "platform": "platforms",
        "framework": "frameworks",
        "concern": "concerns",
        "audience": "audiences",
    }
    dimension = aliases.get(dimension, dimension)
    if dimension not in PROFILE_DIMENSIONS:
        raise ValueError(
            f"unknown profile dimension: {dimension}; "
            f"expected one of: {', '.join(PROFILE_DIMENSIONS)}"
        )
    path = PROFILES_DIR / f"{dimension}.json"
    if not path.is_file():
        raise ValueError(f"profile file not found: {path}")
    return json.loads(path.read_text(encoding="utf-8"))


def load_profiles() -> dict[str, list]:
    return {dimension: load_profile(dimension) for dimension in PROFILE_DIMENSIONS}


def load_type(doc_id: str) -> dict:
    path = TYPES_DIR / f"{doc_id}.json"
    if not path.is_file():
        raise ValueError(f"unknown document type id: {doc_id}")
    return json.loads(path.read_text(encoding="utf-8"))


def load_types(doc_ids: list[str]) -> list[dict]:
    return [load_type(doc_id) for doc_id in doc_ids]


def load_all_types() -> list[dict]:
    index = load_index()
    return [load_type(row["id"]) for row in index["document_types"]]


def as_legacy_catalog() -> dict:
    """Reconstruct the pre-split monolith shape for callers that need the full graph."""
    index = load_index()
    tiers = index["tiers"]
    if isinstance(tiers, dict):
        tier_list = [{"id": tid, "order": meta["order"]} for tid, meta in tiers.items()]
    else:
        tier_list = list(tiers)
    return {
        "$schema": "catalog-schema.json",
        "version": index["version"],
        "tiers": tier_list,
        "profiles": load_profiles(),
        "groups": index.get("groups", []),
        "capabilities": index.get("capabilities", []),
        "documents": load_all_types(),
        "cue_hints": index.get("cue_hints", []),
    }


def tier_rows(tier: str) -> list[dict]:
    index = load_index()
    tiers = index["tiers"]
    if isinstance(tiers, dict):
        if tier not in tiers:
            raise ValueError(f"unknown tier: {tier}")
    else:
        if tier not in {item["id"] for item in tiers}:
            raise ValueError(f"unknown tier: {tier}")
    return [row for row in index["document_types"] if row["tier"] == tier]


def merged_record(doc_id: str) -> dict:
    index = load_index()
    row = next((r for r in index["document_types"] if r["id"] == doc_id), None)
    if row is None:
        raise ValueError(f"unknown document type id: {doc_id}")
    detail = load_type(doc_id)
    return {**detail, "tier": row["tier"], "index_path": row["path"]}


def _normalize_profile_ids(
    dimension: str,
    values: list[str],
    profiles: dict[str, list],
) -> list[str]:
    aliases: dict[str, str] = {}
    for definition in profiles[dimension]:
        aliases[definition["id"]] = definition["id"]
        for alias in definition.get("aliases", []):
            aliases[alias] = definition["id"]
    resolved = []
    for value in values:
        if value not in aliases:
            raise ValueError(f"unknown {dimension[:-1]}: {value}")
        canonical = aliases[value]
        if canonical not in resolved:
            resolved.append(canonical)
    return resolved


def applicable(
    *,
    shape: list[str] | None = None,
    platform: list[str] | None = None,
    framework: list[str] | None = None,
    concern: list[str] | None = None,
    audience: list[str] | None = None,
    tier: str | None = None,
    include_dynamic: bool = False,
) -> list[str]:
    """Return document-type ids applicable for the given profile/tier selection."""
    index = load_index()
    profiles = load_profiles()
    selected = {
        "shapes": _normalize_profile_ids("shapes", shape or [], profiles),
        "platforms": _normalize_profile_ids("platforms", platform or [], profiles),
        "frameworks": _normalize_profile_ids("frameworks", framework or [], profiles),
        "concerns": _normalize_profile_ids("concerns", concern or [], profiles),
        "audiences": _normalize_profile_ids("audiences", audience or [], profiles),
    }
    tiers = index["tiers"]
    if isinstance(tiers, dict):
        ranks = {tid: meta["order"] for tid, meta in tiers.items()}
    else:
        ranks = {item["id"]: item["order"] for item in tiers}
    if tier is None:
        tier = "diligence"
    if tier not in ranks:
        raise ValueError(f"unknown tier: {tier}")
    tier_rank = ranks[tier]

    ids: list[str] = []
    for row in index["document_types"]:
        detail = load_type(row["id"])
        rule = detail["selection"]
        if rule["mode"] == "dynamic" and not include_dynamic:
            continue
        if ranks[rule["min_tier"]] > tier_rank:
            continue
        selectors = rule.get("selectors", {})
        has_selectors = any(selectors.values())
        if has_selectors:
            matched = False
            for dimension, values in selectors.items():
                if dimension == "frameworks":
                    continue  # frameworks must not select documents
                selected_values = selected.get(dimension, [])
                if any(value in selected_values for value in values):
                    matched = True
                    break
            if not matched:
                continue
        ids.append(detail["id"])
    return ids


def validate() -> list[str]:
    errors: list[str] = []
    try:
        index = load_index()
    except ValueError as exc:
        return [str(exc)]
    if index.get("version") != "2.4.0":
        errors.append(f"catalog version must be 2.4.0, got {index.get('version')}")
    for key in ("tiers", "groups", "capabilities", "document_types"):
        if key not in index:
            errors.append(f"index.json missing {key}")
    tiers = index.get("tiers", {})
    if isinstance(tiers, dict):
        tier_ids = set(tiers)
    else:
        tier_ids = {item["id"] for item in tiers}

    profiles: dict[str, list] = {}
    profile_ids: dict[str, set[str]] = {}
    for dimension in PROFILE_DIMENSIONS:
        path = PROFILES_DIR / f"{dimension}.json"
        if not path.is_file():
            errors.append(f"missing profile file: {path.name}")
            profiles[dimension] = []
            profile_ids[dimension] = set()
            continue
        definitions = json.loads(path.read_text(encoding="utf-8"))
        profiles[dimension] = definitions
        ids = {item.get("id") for item in definitions}
        if len(ids) != len(definitions):
            errors.append(f"{dimension}: duplicate profile id")
        profile_ids[dimension] = ids
        names: dict[str, str] = {}
        for item in definitions:
            identifier = item.get("id")
            for name in [identifier, *item.get("aliases", [])]:
                if not isinstance(name, str) or re.fullmatch(r"[a-z0-9][a-z0-9-]*", name) is None:
                    errors.append(f"{dimension}: invalid profile name {name}")
                    continue
                if name in names:
                    errors.append(
                        f"{dimension}: profile name collision {name} "
                        f"between {names[name]} and {identifier}"
                    )
                names[name] = identifier

    index_ids = [row["id"] for row in index.get("document_types", [])]
    if len(index_ids) != len(set(index_ids)):
        errors.append("index.json has duplicate document type ids")
    type_files = {path.stem for path in TYPES_DIR.glob("*.json")}
    for doc_id in index_ids:
        if doc_id not in type_files:
            errors.append(f"index references missing type file: {doc_id}.json")
    for stem in sorted(type_files):
        if stem not in index_ids:
            errors.append(f"orphan type file not in index: {stem}.json")

    groups = set(index.get("groups", []))
    capabilities = set(index.get("capabilities", []))
    static_ids: set[str] = set()
    static_paths: set[str] = set()
    dynamic_types: set[str] = set()
    contract_types: set[str] = set()
    if CONTRACTS_DIR.is_dir():
        contract_types = {
            path.stem for path in CONTRACTS_DIR.glob("*.md") if path.name != "README.md"
        }

    for doc_id in index_ids:
        try:
            doc = load_type(doc_id)
        except ValueError as exc:
            errors.append(str(exc))
            continue
        missing = sorted(REQUIRED_DOC_FIELDS - set(doc))
        if missing:
            errors.append(f"{doc_id}: missing fields: {', '.join(missing)}")
            continue
        if doc["id"] != doc_id:
            errors.append(f"{doc_id}: id field mismatch ({doc['id']})")
        selection = doc.get("selection", {})
        if doc["group"] not in groups:
            errors.append(f"{doc_id}: unknown group {doc['group']}")
        if selection.get("min_tier") not in tier_ids:
            errors.append(f"{doc_id}: unknown tier {selection.get('min_tier')}")
        selectors = selection.get("selectors", {})
        if selectors.get("frameworks"):
            errors.append(f"{doc_id}: frameworks may tailor evidence but must not select documents")
        for dimension, values in selectors.items():
            if dimension not in profile_ids:
                errors.append(f"{doc_id}: unknown selector dimension {dimension}")
                continue
            for value in values:
                if value not in profile_ids[dimension]:
                    errors.append(f"{doc_id}: unknown {dimension} selector {value}")
        for requirement in doc.get("requires", []):
            if requirement not in capabilities:
                errors.append(f"{doc_id}: unknown requirement {requirement}")
        if not isinstance(doc.get("write_order"), int):
            errors.append(f"{doc_id}: write_order must be an integer")
        form = doc.get("dominant_form")
        if form not in ALLOWED_DOMINANT_FORMS:
            errors.append(f"{doc_id}: invalid dominant_form {form!r}")
        if contract_types and doc["type"] not in contract_types:
            errors.append(f"{doc_id}: document type missing from catalog-contracts/")
        template = SKILL_ROOT / "assets" / "templates" / doc["scaffold_template"]
        if not template.is_file():
            errors.append(f"{doc_id}: missing template {doc['scaffold_template']}")
        instruction = doc.get("instruction_file")
        if instruction and not (SKILL_ROOT / "instructions" / instruction).is_file():
            errors.append(f"{doc_id}: missing instruction {instruction}")
        if selection.get("mode") == "static":
            if doc["id"] in static_ids:
                errors.append(f"duplicate static id: {doc['id']}")
            if doc["path"] in static_paths:
                errors.append(f"duplicate static path: {doc['path']}")
            static_ids.add(doc["id"])
            static_paths.add(doc["path"])
        elif selection.get("mode") == "dynamic":
            if doc["type"] in dynamic_types:
                errors.append(f"duplicate dynamic type: {doc['type']}")
            dynamic_types.add(doc["type"])
        else:
            errors.append(f"{doc_id}: selection.mode must be static or dynamic")

    # §0.10: infrastructure-platform must carry widened signals + aliases
    infra = next(
        (item for item in profiles.get("shapes", []) if item.get("id") == "infrastructure-platform"),
        None,
    )
    if infra is None:
        errors.append("shapes: infrastructure-platform missing")
    else:
        patterns = {s.get("pattern") for s in infra.get("signals", [])}
        for required in ("ansible.cfg", "**/kustomization.yaml", "**/playbook*.yml"):
            if required not in patterns:
                errors.append(f"infrastructure-platform missing signal pattern {required}")
        aliases = set(infra.get("aliases", []))
        for alias in ("deployment-config", "iac"):
            if alias not in aliases:
                errors.append(f"infrastructure-platform missing alias {alias}")
    return errors


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--tier", choices=["spine", "diligence", "portfolio"])
    parser.add_argument("--id", dest="doc_id")
    parser.add_argument("--ids", help="Comma-separated document type ids")
    parser.add_argument(
        "--profile",
        help="Profile dimension: shapes|platforms|frameworks|concerns|audiences",
    )
    parser.add_argument("--applicable", action="store_true")
    parser.add_argument("--shape", action="append", default=[])
    parser.add_argument("--platform", action="append", default=[])
    parser.add_argument("--framework", action="append", default=[])
    parser.add_argument("--concern", action="append", default=[])
    parser.add_argument("--audience", action="append", default=[])
    parser.add_argument(
        "--applicable-tier",
        dest="applicable_tier",
        choices=["spine", "diligence", "portfolio"],
        default="diligence",
    )
    parser.add_argument("--include-dynamic", action="store_true")
    parser.add_argument("--legacy", action="store_true", help="Emit reconstructed monolith JSON")
    parser.add_argument("--validate", action="store_true")
    args = parser.parse_args(argv)

    modes = sum(
        bool(x)
        for x in (
            args.tier,
            args.doc_id,
            args.ids,
            args.profile,
            args.applicable,
            args.validate,
            args.legacy,
        )
    )
    if modes != 1:
        return fail(
            "specify exactly one of --tier, --id, --ids, --profile, "
            "--applicable, --legacy, --validate",
            2,
        )

    try:
        if args.validate:
            errors = validate()
            if errors:
                for error in errors:
                    print(f"error: {error}", file=sys.stderr)
                print(f"{len(errors)} validation error(s)", file=sys.stderr)
                return 1
            print("catalog ok")
            return 0
        if args.legacy:
            print(dump_json(as_legacy_catalog()), end="")
            return 0
        if args.tier:
            print(dump_json(tier_rows(args.tier)), end="")
            return 0
        if args.doc_id:
            print(dump_json(merged_record(args.doc_id)), end="")
            return 0
        if args.ids:
            ids = [part.strip() for part in args.ids.split(",") if part.strip()]
            print(dump_json([merged_record(doc_id) for doc_id in ids]), end="")
            return 0
        if args.profile:
            print(dump_json(load_profile(args.profile)), end="")
            return 0
        if args.applicable:
            ids = applicable(
                shape=args.shape,
                platform=args.platform,
                framework=args.framework,
                concern=args.concern,
                audience=args.audience,
                tier=args.applicable_tier,
                include_dynamic=args.include_dynamic,
            )
            print(dump_json(ids), end="")
            return 0
    except ValueError as exc:
        return fail(str(exc), 2)
    return 2


if __name__ == "__main__":
    raise SystemExit(main())
